### PR TITLE
Add retry functionality to Redis cache

### PR DESCRIPTION
**Pull request by svix-daniel**
Begins to add retry functionality to mitigate intermittent Redis failures.

### CODE REVIEW

File: server/svix-server/src/core/cache/mod.rs [hunk 1]
```diff
@@ -108,38 +110,69 @@ impl Cache {
...
+                self.get_raw(key.as_ref().as_bytes())
+                    .await?
+                    .map(|x| {
+                        String::from_utf8(x)
+                            .map_err(|e| e.into())
+                            .and_then(|json| serde_json::from_str(&json).map_err(|e| e.into()))
+                    })
+                    .transpose()
+            },
+            |e| self.should_retry(e),
```
Comment (tasn) [inline server/svix-server/src/core/cache/mod.rs:138]:
```suggestion
            self.should_retry,
```

File: server/svix-server/src/core/run_with_retries.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    should_retry: impl Fn(&E) -> bool,
+    retry_schedule: &[Duration],
+) -> Result<T, E> {
+    let mut num_attempts = 0;
+
+    loop {
+        let fun_ret = fun().await;
+        match fun_ret {
+            Ok(_) => break fun_ret,
+            Err(ref e) => {
```
Comment (tasn) [inline server/svix-server/src/core/run_with_retries.rs:14]: Add a `tracing:warn` when we are retrying.

File: server/svix-server/src/core/run_with_retries.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    for duration in retry_schedule {
+        let fun_ret = fun().await;
+        match fun_ret {
+            Ok(_) => return fun_ret,
+            Err(ref e) => {
+                if should_retry(e) {
+                    tracing::warn!("Retrying after error {}", e);
+                    tokio::time::sleep(*duration).await;
+                } else {
+                    tracing::warn!("Unrecoverable error {}", e);
```
Comment (tasn) [inline server/svix-server/src/core/run_with_retries.rs:22]: No need...

File: server/svix-server/src/core/run_with_retries.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                    return fun_ret;
+                }
+            }
+        }
+    }
+
+    tracing::warn!("Exhausted retry schedule");
```
Comment (tasn) [inline server/svix-server/src/core/run_with_retries.rs:29]: No need.

File: server/svix-server/src/core/run_with_retries.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    fun().await
```
Comment (tasn) [inline server/svix-server/src/core/run_with_retries.rs:30]: Why make another call here?

File: server/svix-server/src/core/cache/mod.rs [hunk 6]
```diff
@@ -108,38 +110,69 @@ impl Cache {
...
+    Duration::from_millis(20),
+    Duration::from_millis(40),
+];
+
 #[async_trait]
 #[enum_dispatch(Cache)]
 pub trait CacheBehavior: Sync + Send {
+    // TODO: Reduce retry-related duplication
+    // TODO: Test retries
+    // TODO: Retry schedule constant
```
Comment (tasn) [inline server/svix-server/src/core/cache/mod.rs:124]: This is not a place for todo, remove it.

**Review by svix-dylan — APPROVED** [review]

**Pull request merged**